pallets/providers/src/utils.rs

```diff
@@ -208,16 +215,156 @@ where
+                Self::do_msp_sign_up(who, &msp_info, request_block)?;
+            }
+            StorageProvider::BackupStorageProvider(bsp_info) => {
+                Self::do_bsp_sign_up(who, &bsp_info, request_block)?;
+            }
+        }
+
+        Ok(())
+    }
+
+    /// This function holds the logic that confirms the sign up of a user as a Main Storage Provider
+    /// It updates the storage to add the new Main Storage Provider, increments the counter of Main Storage Providers,
+    /// and removes the sign up request from the SignUpRequests mapping
+    pub fn do_msp_sign_up(
+        who: &T::AccountId,
+        msp_info: &MainStorageProvider<T>,
+        request_block: BlockNumberFor<T>,
+    ) -> DispatchResult {
+        // Check that, by registering this Main Storage Provider, we are not exceeding the maximum number of Main Storage Providers
+        let new_amount_of_msps = MspCount::<T>::get()
+            .checked_add(&T::SpCount::one())
+            .ok_or(DispatchError::Arithmetic(ArithmeticError::Overflow))?;
+        ensure!(
+            new_amount_of_msps <= T::MaxMsps::get(),
+            Error::<T>::MaxMspsReached
+        );
+
+        // Get the MainStorageProviderId by using the AccountId as the seed for a random generator
+        let (msp_id, block_number_when_random) =
+            T::ProvidersRandomness::random(who.encode().as_ref());
+
+        // Check that the maximum block number after which the randomness is invalid is greater than or equal to the block number when the
+        // request was made to ensure that the randomness was not known when the request was made
+        ensure!(
+            block_number_when_random >= request_block,
+            Error::<T>::RandomnessNotValidYet
+        );
+
+        // Check that the current block number is not greater than the block number when the request was made plus the maximum amount of
+        // blocks that we allow the user to wait for valid randomness (should be at least more than an epoch if using BABE's RandomnessFromOneEpochAgo)
+        // We do this to ensure that a user cannot wait indefinitely for randomness that suits them
+        ensure!(
+            frame_system::Pallet::<T>::block_number()
+                < request_block + T::MaxBlocksForRandomness::get(),
+            Error::<T>::SignUpRequestExpired
+        );
+
+        // Insert the MainStorageProviderId into the mapping
+        AccountIdToMainStorageProviderId::<T>::insert(who, msp_id);
+
+        // Save the MainStorageProvider information in storage
+        MainStorageProviders::<T>::insert(&msp_id, msp_info);
+
+        // Increment the counter of Main Storage Providers registered
+        MspCount::<T>::set(new_amount_of_msps);
+
+        // Remove the sign up request from the SignUpRequests mapping
+        SignUpRequests::<T>::remove(who);
+
+        // Emit the corresponding event
+        Self::deposit_event(Event::<T>::MspSignUpSuccess {
```

Comment:
In the convention we're using, events are emitted in the `lib.rs` file. I'm not against emitting them here though, but we should have a discussion about it and come to an agreement.

Reply:
I see why you do this. You have one single extrinsic that can emit one or the other event. Maybe this is a good argument to emit events here...

Reply:
Yep! Lets discuss it tomorrow, but that's exactly the reason why. I could also split it into two extrinsics and can then emit events where we usually do. It wouldn't be as user-friendly but it's not that bad either

Reply:
Ok, I think it is fine in this case.